### ISSUE TITLE
Parsing of Walrus Operator 

### ISSUE THREAD

**Issue by rvanlaar** [python/mypy#8053]
Mypy doesn't seem to handle variable type changes that happen with a walrus operator.

What happens here is: the walrus operator makes it a boolean:
```while i:= update(i) > 0```. 

This is what needs to happen to check for `i > 0` and `i` still being an int.
```while (i:= update(i)) > 0 ```. 

The following python code doesn't give any errors with mypy 0.750:

```python
def walrus1(i: int) -> int:
    """Actually returns bool"""
    while i:= i > 0:
       return i 
    return i

def walrus2(i: int) -> int:
    """Works correctly."""
    while (i:= i) > 0:
        return i
    return i

assert walrus1(5) == True
assert walrus1(0) == False
assert walrus2(5) == 5
assert walrus2(0) == 0
```

**Comment by rvanlaar** [comment]
Where in the mypy code can I loop If I were to try to make a pull request?

**Comment by JelleZijlstra** [comment]
I don't know exactly, but #6899 is the PR where I added (limited) support for the walrus operator, so the bug may be in the code added there.

**Comment by rvanlaar** [comment]
It seems that `bool` is a covariant of `int`.

In the following example:  `x` and  is alright, `y` and `z` aren't.

```python
x: int
x = True

y: bool
y = 5

z: str
z = True
```

**Comment by JelleZijlstra** [comment]
Right, `bool` is a subclass of `int`, so mypy sees that the function still returns a kind of int.

**Comment by rvanlaar** [comment]
Which wasn't what I expected. Even with years of python experience. 

Could the table on https://mypy.readthedocs.io/en/latest/builtin_types.html be updated? 
That's were I looked for an explanation.

I would expect that the definition of bool there would be a subtype of int.

**Comment by JukkaL** [comment]
@rvanlaar Sure, updating the table is good idea. Would like to create a PR (or a separate issue to track it)?

**Comment by rvanlaar** [comment]
Closing this issue, continue with issue #8069 to document it.
